### PR TITLE
Support T-SQL Linked Servers

### PR DESCRIPTION
**Pull request by thephantomthief**
### Description

This commit adds the support for T-SQL linked server objects through
PostgreSQL foreign data wrappers (FDWs). Currently, only support for
linked servers for remote SQL Servers is added. Thus, to add linked
servers of this type, the tds_fdw extension needs to be installed.

Also, added the support for T-SQL sp_addlinkedserver stored procedure to
add a linked server in Babelfish and sys.servers system view to show
list of linked servers.

sp_addlinkedserver stored procedure has been implemented such that
internally it uses PG's CREATE FOREIGN SERVER logic and so, a linked
server added is internally a PG foreign server. sys.servers system view
uses the pg_foreign_server catalog internally.

Task: BABEL-861
Signed-off-by: Sharu Goel <goelshar@amazon.com>

### Test Scenarios Covered ###
* **Use case based -** Yes

* **Boundary conditions -** N/A

* **Arbitrary inputs -** Yes

* **Negative test cases -** Yes

* **Minor version upgrade tests -** Yes

* **Major version upgrade tests -** Yes

* **Performance tests -** N/A

* **Tooling impact -** Yes. Some tools require sp_addlinkedserver to be in master.dbo schema, so incorporated the same.

* **Client tests -** N/A

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: contrib/babelfishpg_tsql/src/procedures.c [hunk 1]
```diff
@@ -2070,3 +2076,151 @@ gen_sp_droprolemember_subcmds(const char *user, const char *member)
...
+}
+
+/*
+ * Internal function to create the following procedures related to T-SQL linked
+ * servers in master.dbo schema:
+ *   - sp_addlinkedserver
+ * Some applications invoke this referencing master.dbo.<one of the above stored procedures>
+ */
+Datum
+create_linked_server_procs_in_master_dbo_internal(PG_FUNCTION_ARGS)
```
Comment (KushaalShroff) [inline contrib/babelfishpg_tsql/src/procedures.c]: Can you please explain why we have created a C type function to create another function?
>  CALL sys.create_linked_server_procs_in_master_dbo();
  ALTER PROCEDURE master_dbo.sp_addlinkedserver OWNER TO sysadmin;
  DROP PROCEDURE sys.create_linked_server_procs_in_master_dbo;

The above logic seems counter intuitive and kind of confusing to understand the need for sys.create_linked_server_procs_in_master_dbo

File: contrib/babelfishpg_tsql/src/procedures.c [hunk 2]
```diff
@@ -2070,3 +2076,151 @@ gen_sp_droprolemember_subcmds(const char *user, const char *member)
...
+						"IN \"@srvproduct\" sys.nvarchar(128) DEFAULT NULL,"
+						"IN \"@provider\" sys.nvarchar(128) DEFAULT 'SQLNCLI',"
+						"IN \"@datasrc\" sys.nvarchar(4000) DEFAULT NULL,"
+						"IN \"@location\" sys.nvarchar(4000) DEFAULT NULL,"
+						"IN \"@provstr\" sys.nvarchar(4000) DEFAULT NULL,"
+						"IN \"@catalog\" sys.sysname DEFAULT NULL)"
+						"AS \'babelfishpg_tsql\', \'sp_addlinkedserver_internal\'"
+					"LANGUAGE C";
+
+	const char  *dbo_scm = get_dbo_schema_name("master");
```
Comment (KushaalShroff) [inline contrib/babelfishpg_tsql/src/procedures.c]: In ownership.sql you have blindly used master_dbo, where as here we are relying on an API, can you please explain why we need that? If so why arent we doing something similar in .sql?

File: contrib/babelfishpg_tsql/src/procedures.c [hunk 3]
```diff
@@ -2070,3 +2076,151 @@ gen_sp_droprolemember_subcmds(const char *user, const char *member)
...
+{
+	char *linked_server = PG_ARGISNULL(0) ? NULL : text_to_cstring(PG_GETARG_TEXT_P(0));
+	char *srv_product = PG_ARGISNULL(1) ? "" : text_to_cstring(PG_GETARG_TEXT_P(1));
+	char *provider = PG_ARGISNULL(2) ? "" : text_to_cstring(PG_GETARG_TEXT_P(2));
+	char *data_src = PG_ARGISNULL(3) ? "" : text_to_cstring(PG_GETARG_TEXT_P(3));
+	char *provstr = PG_ARGISNULL(5) ? NULL : text_to_cstring(PG_GETARG_TEXT_P(5));
+	char *catalog = PG_ARGISNULL(6) ? NULL : text_to_cstring(PG_GETARG_TEXT_P(6));
+
+	CreateForeignServerStmt *stmt = makeNode(CreateForeignServerStmt);
+	List *options = NIL;
```
Comment (KushaalShroff) [inline contrib/babelfishpg_tsql/src/procedures.c]: We should free all the pointers/lists

File: contrib/babelfishpg_tsql/src/procedures.c [hunk 4]
```diff
@@ -2070,3 +2076,151 @@ gen_sp_droprolemember_subcmds(const char *user, const char *member)
 	rewrite_object_refs(stmt);
 	return res;
 }
+
+static void
+ValidateLinkedServerDataSource(char* data_src)
```
Comment (KushaalShroff) [inline contrib/babelfishpg_tsql/src/procedures.c]: Do we really need a function for this?
We can use a macro instead. Moreover, we are using it only once so I dont see a point.

File: .github/composite-actions/setup-base-version/action.yml [hunk 5]
```diff
@@ -43,6 +43,18 @@ runs:
...
+      run: |
+        cd ..
+        export TDS_FDW_VERSION="2.0.3"
+        sudo apt-get install wget
+        wget https://github.com/tds-fdw/tds_fdw/archive/v${TDS_FDW_VERSION}.tar.gz
+        tar -xvzf v${TDS_FDW_VERSION}.tar.gz
+        cd tds_fdw-${TDS_FDW_VERSION}/
+        make USE_PGXS=1 PG_CONFIG=~/${{ inputs.install_dir }}/bin/pg_config
+        sudo make USE_PGXS=1 PG_CONFIG=~/${{ inputs.install_dir }}/bin/pg_config install
+      shell: bash
```
Comment (Deepesh125) [inline .github/composite-actions/setup-base-version/action.yml:56]: are these steps same as composite-actions/build-tds_fdw-extension? if so can we utilise the same here? and possibly at other places?

File: contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.0.0--3.1.0.sql [hunk 6]
```diff
@@ -210,6 +210,81 @@ CREATE OR REPLACE FUNCTION sys.degrees(IN arg1 TINYINT)
...
+
+ALTER PROCEDURE master_dbo.sp_addlinkedserver OWNER TO sysadmin;
+
+CREATE OR REPLACE VIEW sys.servers
+AS
+SELECT
+  CAST(f.oid as int) AS server_id,
+  CAST(f.srvname as sys.sysname) AS name,
+  CAST('' as sys.sysname) AS product,
+  CAST('tds_fdw' as sys.sysname) AS provider,
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.0.0--3.1.0.sql:265]: From T-sql documentation - `The value in provider_string is always NULL unless the caller has the ALTER ANY LINKED SERVER permission.`

File: contrib/babelfishpg_tsql/src/procedures.c [hunk 7]
```diff
@@ -2070,3 +2075,148 @@ gen_sp_droprolemember_subcmds(const char *user, const char *member)
...
+			/* we ignore provider string in any case */
+			provstr_warning = true;
+		}
+	}
+
+	initStringInfo(&query);
+
+	appendStringInfo(&query, "CREATE SERVER \"%s\" FOREIGN DATA WRAPPER tds_fdw ", linked_server);
+
+	/* Add the relevant options */
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/src/procedures.c:2181]: Please add comment with more info, give an example how options would look like with final query string?

File: test/python/expected/upgrade_validation/expected_dependency.out [hunk 8]
```diff
@@ -981,6 +981,7 @@ View sys.master_files
 View sys.pg_namespace_ext
 View sys.registered_search_property_lists
 View sys.selective_xml_index_paths
+View sys.servers
```
Comment (Deepesh125) [inline test/python/expected/upgrade_validation/expected_dependency.out]: let's add dependency test cases for sys.servers

**Review by Deepesh125 — CHANGES_REQUESTED** [review]

File: contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.0.0--3.1.0.sql (hunk 6 above)
Comment (thephantomthief) [reply contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.0.0--3.1.0.sql:265]: provider is different from provider string

File: .github/composite-actions/setup-base-version/action.yml (hunk 5 above)
Comment (thephantomthief) [reply .github/composite-actions/setup-base-version/action.yml:56]: at places where we checkout older branches, we cannot reuse the composite action

**Review by Deepesh125 — APPROVED** [review]

**Review by kuntalghosh — APPROVED** [review]

**Pull request merged**